File: loopx/control_plane/agents/capability_gate.ts
/** Read policy only: requirements are not enablement, credentials or write authority. */
import type {JsonObject} from "../effect_program.ts";
import {requireJsonObject, requireStringArray, requireInteger, requireNonEmptyString, requireBoolean, optionalNonEmptyString} from "../runtime_decode.ts";

const DEFAULT_AVAILABLE = ["shell", "filesystem_read", "filesystem_write"];
const OWNER_HELD = new Set(["credentials", "production_access"]);
export const OBSERVABLE_RUNTIME_CAPABILITIES: ReadonlySet<string> = new Set(["benchmark_runner", "network", "external_evidence_poll", "worker_bridge", "cli_bridge"]);
type CapabilityAction = "run" | "ask_owner" | "repair_bridge";
type ResolutionOwner = "user" | "agent";
interface Requirement {required: string[]; targets: string[]}
interface Candidate extends Requirement {payload: JsonObject; rank: number[]; ordinal: number}
interface Binding {
  owner: ResolutionOwner; action: "provide_or_authorize" | "repair_bridge"; capability: string;
  priority: string; primary_blocked_todo_id: string | null; blocked_todo_ids: string[];
}
const unique = (values: readonly string[]) => [...new Set(values)];
export const availableCapabilities = (values: readonly string[]): string[] => unique([...DEFAULT_AVAILABLE, ...values]);

/** A repair output is not its own prerequisite; this does not grant that output. */
export function missingRequiredCapabilities(required: readonly string[], targets: readonly string[], available: readonly string[]): string[] {
  const present = new Set([...availableCapabilities(available), ...targets]);
  return required.filter(capability => !present.has(capability));
}
const action = (missing: readonly string[]): CapabilityAction => !missing.length ? "run" :
  missing.some(capability => OWNER_HELD.has(capability)) ? "ask_owner" : "repair_bridge";
function requirement(value: unknown): Requirement {
  const row = requireJsonObject(value, "capability requirement");
  return {required: requireStringArray(row.required, "required"), targets: requireStringArray(row.targets, "targets")};
}
function list(value: unknown): unknown[] {
  if (!Array.isArray(value)) throw new TypeError("capability rows must be an array");
  return value;
}
const values = (rows: readonly JsonObject[], key: string) => unique(rows.flatMap(row => row[key] as string[] ?? []));
const priority = (row: JsonObject): string => /^(P[0-2])/.exec(String(row.priority ?? "").trim().toUpperCase())?.[1] ?? "P1";

function bindings(blocked: readonly JsonObject[]): Binding[] {
  const result = new Map<string, Binding>();
  for (const row of blocked) for (const capability of row.missing_capabilities as string[]) {
    const owner: ResolutionOwner = OWNER_HELD.has(capability) ? "user" : "agent";
    const id = String(row.todo_id ?? "").trim() || null;
    const key = `${owner}:${capability}`;
    let binding = result.get(key);
    if (!binding) {
      binding = {owner, action: owner === "user" ? "provide_or_authorize" : "repair_bridge", capability,
        priority: priority(row), primary_blocked_todo_id: id, blocked_todo_ids: []};
      result.set(key, binding);
    } else if (priority(row) < binding.priority) {
      // Display/source order is not the urgency of a shared repair obligation.
      binding.priority = priority(row); binding.primary_blocked_todo_id = id;
    }
    if (id && !binding.blocked_todo_ids.includes(id)) binding.blocked_todo_ids.push(id);
  }
  return [...result.values()];
}

export function projectCapabilityGate(request: JsonObject): JsonObject | null {
  const available = availableCapabilities(requireStringArray(request.available, "available"));
  const source = requireNonEmptyString(request.source, "source");
  const candidates: Candidate[] = list(request.candidates).map((value, ordinal) => {
    const row = requireJsonObject(value, "capability candidate");
    const rank = list(row.rank).map(value => requireInteger(value, "rank"));
    if (rank.length !== 6) throw new TypeError("capability candidate rank must have six dimensions");
    return {...requirement(row), payload: requireJsonObject(row.payload, "payload"), rank, ordinal};
  });
  const blocked: JsonObject[] = [], runnable: {payload: JsonObject; rank: number[]; ordinal: number}[] = [];
  let sawRequirement = false;
  for (const candidate of candidates) {
    const {required, targets} = candidate;
    sawRequirement ||= !!(required.length || targets.length);
    const missing = missingRequiredCapabilities(required, targets, available);
    const missingTargets = missing.length ? [] : targets.filter(capability => !available.includes(capability));
    const repair = missingTargets.some(capability => OBSERVABLE_RUNTIME_CAPABILITIES.has(capability));
    const payload = {...candidate.payload, required_capabilities: required,
      ...(targets.length ? {target_capabilities: targets} : {}), missing_capabilities: missing,
      capability_action: repair ? "repair_bridge" : action(missing),
      ...(missingTargets.length ? {missing_target_capabilities: missingTargets} : {}),
      ...(repair ? {capability_repair_mode: true} : {})};
    if (missing.length) blocked.push(payload);
    else {
      const rank = [...candidate.rank]; rank[4] = repair ? 0 : 1;
      runnable.push({payload, rank, ordinal: candidate.ordinal});
    }
  }
  if (!sawRequirement && !blocked.length) return null;
  const policy = request.candidate_order_policy == null ? null : requireNonEmptyString(request.candidate_order_policy, "candidate_order_policy");
  if (policy) runnable.sort((a, b) => {
    for (let i = 0; i < a.rank.length; i++) if (a.rank[i] !== b.rank[i]) return a.rank[i]! - b.rank[i]!;
    return a.ordinal - b.ordinal;
  });
  const rows = runnable.map(row => row.payload), resolution = bindings(blocked);
  const forOwner = (owner: ResolutionOwner) => resolution.filter(binding => binding.owner === owner).map(binding => binding.capability);
  const ownerAction = resolution.filter(binding => binding.owner === "user").map(binding =>
    binding.primary_blocked_todo_id ? `${binding.capability} for ${binding.primary_blocked_todo_id}` : binding.capability);
  const common = {schema_version: "capability_gate_v0", source, available,
    runnable_count: rows.length, runnable_candidates: rows, blocked_candidates: blocked,
    owner_missing: forOwner("user"), unsupported_missing: [], resolution_bindings: resolution,
    owner_action: ownerAction.length ? `provide or authorize the missing owner-held capability: ${ownerAction.join(", ")}` : null};
  if (rows.length) return {...common, required: values(rows, "required_capabilities"), missing: [], action: "run",
    decision_owner: "agent", selection_policy: "agent_steering_audit_over_runnable_candidates",
    candidate_order_policy: policy ?? "projection_order", blocked_missing: values(blocked, "missing_capabilities"),
    repair_missing: unique([...values(rows.filter(row => row.capability_repair_mode === true), "missing_target_capabilities"), ...forOwner("agent")]),
    repair_candidate_count: rows.filter(row => row.capability_repair_mode === true).length,
    reason: "capability gate projected runnable candidate set; agent chooses the actual todo"};
  const missing = values(blocked, "missing_capabilities"), ownerMissing = forOwner("user"), repairMissing = forOwner("agent");
  return {...common, required: values(blocked, "required_capabilities"), missing, action: action(missing),
    decision_owner: ownerMissing.length ? "user" : "agent", repair_missing: repairMissing,
    resolution_steps: [
      ...(ownerMissing.length ? [{owner: "user", action: "provide_or_authorize", capabilities: ownerMissing}] : []),
      ...(repairMissing.length ? [{owner: "agent", action: "repair_bridge", capabilities: repairMissing}] : []),
    ], selection_policy: "no_runnable_candidate", blocks_delivery: true,
    reason: "all visible executable todo candidates require unavailable capabilities"};
}

/** Host-local read plan: never writes a grant or changes a committed Turn binding. */
export function projectRuntimeCapabilityReentry(request: JsonObject): JsonObject | null {
  const gate = requireJsonObject(request.gate, "gate");
  const observed = unique(requireStringArray(request.available, "available").filter(c => !OWNER_HELD.has(c)));
  const selectionRequired = requireBoolean(request.selection_required, "selection_required");
  const selectedId = optionalNonEmptyString(request.selected_todo_id, "selected_todo_id");
  const receiptId = optionalNonEmptyString(request.receipt_todo_id, "receipt_todo_id");
  // A receipt remains binding even if an inconsistent advisory flag accompanies it.
  const boundId = receiptId ?? (selectionRequired ? null : selectedId);
  const baseArgs = requireStringArray(request.command_prefix, "command_prefix");
  const schedulerArgs = requireStringArray(request.scheduler_args, "scheduler_args");
  const missing = requireStringArray(gate.repair_missing ?? [], "repair_missing")
    .filter(c => !OWNER_HELD.has(c) && !observed.includes(c));
  const resolutions = list(gate.resolution_bindings ?? []).map(value => {
    const row = requireJsonObject(value, "resolution binding");
    return {
      owner: requireNonEmptyString(row.owner, "owner"),
      capability: requireNonEmptyString(row.capability, "capability"),
      primary: optionalNonEmptyString(row.primary_blocked_todo_id, "primary_blocked_todo_id"),
      ids: requireStringArray(row.blocked_todo_ids ?? [], "blocked_todo_ids"),
    };
  });
  const blocked = list(gate.blocked_candidates ?? []).map(value => {
    const row = requireJsonObject(value, "blocked candidate");
    return {
      id: optionalNonEmptyString(row.todo_id, "todo_id"),
      instruction: optionalNonEmptyString(row.text, "text"),
      action: optionalNonEmptyString(row.action_kind, "action_kind") ?? "unspecified",
      target: optionalNonEmptyString(row.target_key, "target_key"),
      required: requireStringArray(row.required_capabilities ?? [], "required_capabilities"),
    };
  });
  if (!schedulerArgs.length) return null;
  const candidates: JsonObject[] = [];
  for (const capability of unique(missing)) {
    const bindings = resolutions.filter(row => row.owner === "agent" && row.capability === capability);
    const ids = new Set(bindings.flatMap(row => row.ids.length ? row.ids : row.primary ? [row.primary] : []));
    const eligible = blocked.filter(row => row.id && ids.has(row.id) && row.instruction &&
      row.required.includes(capability) && (!boundId || row.id === boundId));
    // Prefer the gate's highest-priority target, not incidental display order.
    const target = eligible.find(row => bindings.some(binding => binding.primary === row.id)) ?? eligible[0];
    if (!target) continue;
    candidates.push({
      capability, verification_required: "successful_real_callsite_observation",
      verification_target: {todo_id: target.id!, action_kind: target.action, instruction: target.instruction!,
        ...(target.target ? {target_ref: target.target} : {})},
      command_argv: [...baseArgs, ...observed.flatMap(c => ["--available-capability", c]),
        "--available-capability", capability, ...schedulerArgs],
    });
  }
  if (!candidates.length) return null;
  return {
    schema_version: "runtime_capability_reentry_v0", state: "verification_required",
    source: "quota_should_run.capability_gate.repair_missing", candidates,
    verification_contract: {scope: "real_task_facing_callsite_for_blocked_todo", ordinary_delivery_allowed: false,
      advancement_checkpoint: false, settles_turn: false,
      on_success: "rerun_quota_in_same_turn_then_continue_if_allowed",
      on_failure: "record_exact_blocker_without_capability_flag"},
    inheritance_contract: {source_invocation: "verified quota should-run reentry",
      propagates_to: ["interaction_contract.cli_channel.next_cli_actions", "quota spend-slot", "quota monitor-poll"],
      session_scoped: false, observation_scope: "host_registry_goal_agent",
      remembers: [...OBSERVABLE_RUNTIME_CAPABILITIES], durable_grant_written: false},
    failure_policy: "Do not add the capability flag when the real callsite check fails; continue the capability repair or record the concrete blocker.",
  };
}

/** Agent negatives override inherited declarations; a fresh explicit observation wins. */
export function projectCapabilityAvailability(request: JsonObject): JsonObject {
  const goal = unique(requireStringArray(request.goal, "goal"));
  const runtime = unique(requireStringArray(request.runtime, "runtime"));
  const state = requireJsonObject(request.agent ?? {}, "agent");
  const remembered = requireStringArray(state.available ?? [], "agent.available")
    .filter(c => OBSERVABLE_RUNTIME_CAPABILITIES.has(c));
  const unavailable = requireStringArray(state.unavailable ?? [], "agent.unavailable")
    .filter(c => OBSERVABLE_RUNTIME_CAPABILITIES.has(c) && !runtime.includes(c));
  const runtimeAvailable = unique([...remembered, ...runtime]).filter(c => !unavailable.includes(c));
  return {goal, agent: remembered, unavailable, invocation: runtime,
    runtime_available: runtimeAvailable,
    effective: unique([...goal, ...runtimeAvailable]).filter(c => !unavailable.includes(c))};
}

export function evaluateCapabilityGate(value: unknown): JsonObject {
  const request = requireJsonObject(value, "capability gate request");
  if (request.schema_version !== "capability_gate_request_v0") throw new TypeError("capability gate request schema mismatch");
  if (request.operation === "availability") return {schema_version: "capability_gate_result_v0", result: projectCapabilityAvailability(request)};
  if (request.operation === "project") return {schema_version: "capability_gate_result_v0", result: projectCapabilityGate(request)};
  if (request.operation === "reentry") return {schema_version: "capability_gate_result_v0", result: projectRuntimeCapabilityReentry(request)};
  if (request.operation === "missing") {
    const available = requireStringArray(request.available, "available");
    return {schema_version: "capability_gate_result_v0", result: list(request.items).map(item => {
      const {required, targets} = requirement(item);
      return missingRequiredCapabilities(required, targets, available);
    })};
  }
  throw new TypeError("unknown capability gate operation");
}

File: loopx/control_plane/agents/capability_gate.py
from __future__ import annotations

from functools import lru_cache
from typing import Any

from ..todos.contract import (
    TODO_TASK_CLASS_ADVANCEMENT,
    normalize_required_capabilities,
    normalize_target_capabilities,
    normalize_todo_claimed_by,
)
from ..todos.todo_semantics import (
    todo_index_rank,
    todo_item_is_actionable_open,
    todo_item_task_class,
    todo_priority_rank,
)
from ..todos.summary_item import compact_todo_summary_item
from .agent_scope import agent_scope_item_claimed_by
from .profile import agent_profile_candidate_rank


CAPABILITY_OWNER_GATE_HINTS = {
    "credentials",
    "production_access",
}


def runtime_capabilities_for_cli_projection(value: Any) -> list[str]:
    """Return observed runtime capabilities, never owner-held authority gates."""

    return [
        capability
        for capability in normalize_required_capabilities(value)
        if capability not in CAPABILITY_OWNER_GATE_HINTS
    ]


def _evaluate(operation: str, **facts: Any) -> Any:
    from ..effect_runtime import effect_runtime_result

    response = effect_runtime_result("agent.capability_gate.evaluate", {
        "schema_version": "capability_gate_request_v0", "operation": operation, **facts,
    })
    if not isinstance(response, dict) or response.get("schema_version") != "capability_gate_result_v0":
        raise TypeError("invalid typed capability gate result")
    return response["result"]


@lru_cache(maxsize=512)
def _missing(required: tuple[str, ...], targets: tuple[str, ...],
             available: tuple[str, ...]) -> tuple[str, ...]:
    # Cache pure normalized requirement values, never Todo identity/state or authority.
    result = _evaluate("missing", items=[{"required": list(required), "targets": list(targets)}],
                       available=list(available))
    return tuple(result[0])


def missing_required_capabilities(item: dict[str, Any], *, available_capabilities: Any) -> list[str]:
    required = tuple(normalize_required_capabilities(item.get("required_capabilities")))
    if not required:
        return []
    return list(_missing(required, tuple(normalize_target_capabilities(item.get("target_capabilities"))),
                         tuple(normalize_required_capabilities(available_capabilities))))


def _capability_item_identity(item: dict[str, Any]) -> tuple[str, str]:
    return (
        str(item.get("todo_id") or ""),
        "" if item.get("todo_id") else str(item.get("text") or "").strip(),
    )


def _agent_lane_candidate_sort_key(
    raw_item: dict[str, Any],
    *,
    agent_id: str | None,
    preferred_todo_ids: set[str] | None = None,
    agent_profile: dict[str, Any] | None = None,
) -> tuple[int, int, int, int, int, int]:
    preferred_todo_ids = preferred_todo_ids or set()
    todo_id = str(raw_item.get("todo_id") or "").strip()
    active_next_rank = 0 if todo_id and todo_id in preferred_todo_ids else 1
    claimed_by = agent_scope_item_claimed_by(raw_item)
    claim_rank = 0 if agent_id and claimed_by == agent_id else 1
    repair_rank = 0 if raw_item.get("capability_repair_mode") is True else 1
    # Durable Next Action is a steering hint inside the selected peer/profile
    # priority bucket, not permission to cross an explicit todo priority boundary.
    return (
        claim_rank,
        agent_profile_candidate_rank(raw_item, agent_profile=agent_profile),
        todo_priority_rank(raw_item),
        active_next_rank,
        repair_rank,
        todo_index_rank(raw_item),
    )


def _select_advancement_candidate_source(
    agent_todo_summary: dict[str, Any],
) -> tuple[list[Any], str]:
    active_next_items = agent_todo_summary.get("active_next_action_executable_items")
    backlog_items = agent_todo_summary.get("executable_backlog_items")
    first_executable_items = agent_todo_summary.get("first_executable_items")
    if isinstance(active_next_items, list) and active_next_items:
        return (
            [
                *active_next_items,
                *(backlog_items if isinstance(backlog_items, list) else []),
            ],
            "agent_todo_summary.active_next_action_executable_items",
        )
    if isinstance(backlog_items, list):
        return backlog_items, "agent_todo_summary.executable_backlog_items"
    if isinstance(first_executable_items, list) and first_executable_items:
        return first_executable_items, "agent_todo_summary.first_executable_items"
    return [], "agent_todo_summary.executable_backlog_items"


def _collect_capability_gate_candidates(
    agent_todo_summary: dict[str, Any],
) -> tuple[list[dict[str, Any]], str]:
    raw_items, source = _select_advancement_candidate_source(agent_todo_summary)
    raw_due_monitor_items = agent_todo_summary.get("monitor_due_items")
    due_monitor_items = (
        raw_due_monitor_items if isinstance(raw_due_monitor_items, list) else []
    )
    raw_blocked_due_monitor_items = agent_todo_summary.get(
        "monitor_capability_blocked_due_items"
    )
    blocked_due_monitor_items = (
        raw_blocked_due_monitor_items
        if isinstance(raw_blocked_due_monitor_items, list)
        else []
    )
    monitor_sources: list[str] = []
    if due_monitor_items:
        monitor_sources.append("agent_todo_summary.monitor_due_items")
    if blocked_due_monitor_items:
        monitor_sources.append(
            "agent_todo_summary.monitor_capability_blocked_due_items"
        )
    if monitor_sources:
        source_parts = [source, *monitor_sources] if raw_items else monitor_sources
        source = "+".join(source_parts)
        raw_items = [*raw_items, *due_monitor_items, *blocked_due_monitor_items]

    deduped_items: list[dict[str, Any]] = []
    seen: set[tuple[str, str]] = set()
    for item in raw_items:
        if not isinstance(item, dict):
            continue
        identity = _capability_item_identity(item)
        if identity in seen:
            continue
        seen.add(identity)
        deduped_items.append(item)

    due_monitor_identities = {
        _capability_item_identity(item)
        for item in [*due_monitor_items, *blocked_due_monitor_items]
        if isinstance(item, dict)
    }
    return (
        [
            item
            for item in deduped_items
            if todo_item_is_actionable_open(item)
            and (
                todo_item_task_class(item) == TODO_TASK_CLASS_ADVANCEMENT
                or _capability_item_identity(item) in due_monitor_identities
            )
        ],
        source,
    )


def build_capability_gate(
    agent_todo_summary: dict[str, Any] | None, *,
    available_capabilities: list[str], agent_identity: dict[str, Any] | None = None,
) -> dict[str, Any] | None:
    if not isinstance(agent_todo_summary, dict):
        return None
    candidates, source = _collect_capability_gate_candidates(agent_todo_summary)
    if not candidates:
        return None
    identity = agent_identity if isinstance(agent_identity, dict) else {}
    agent = normalize_todo_claimed_by(identity.get("agent_id"))
    profile = identity.get("agent_profile")
    profile = profile if isinstance(profile, dict) else None
    policy = ("claim_then_profile_then_priority_then_active_next_then_repair" if profile else
              "claim_then_priority_then_active_next_then_repair") if agent else None
    return _evaluate("project", source=source,
                     available=normalize_required_capabilities(available_capabilities),
                     candidate_order_policy=policy, candidates=[{
                         "payload": compact_todo_summary_item(item, text=str(item.get("text") or "").strip()),
                         "required": normalize_required_capabilities(item.get("required_capabilities")),
                         "targets": normalize_target_capabilities(item.get("target_capabilities")),
                         "rank": list(_agent_lane_candidate_sort_key(item, agent_id=agent, agent_profile=profile)),
                     } for item in candidates])
